File: Python/Lesson_172_n_queens.py
# =============================================================
# MIT License | @analyticswithharry2026
# GitHub  : https://github.com/analyticswithharry
# YouTube : Analytics with Harry
# =============================================================
# Lesson     : 172 -- N Queens
# Category   : Backtracking
# Difficulty : Hard
# Study Plan : Day 86
# =============================================================
#
# QUESTION:
#   Place n queens on n×n board so none attack each other; return the count of distinct solutions.
# =============================================================
def totalNQueens(n):
    cnt=[0]; cols=set(); d1=set(); d2=set()
    def bt(r):
        if r==n: cnt[0]+=1; return
        for c in range(n):
            if c in cols or r-c in d1 or r+c in d2: continue
            cols.add(c); d1.add(r-c); d2.add(r+c)
            bt(r+1)
            cols.remove(c); d1.remove(r-c); d2.remove(r+c)
    bt(0); return cnt[0]

if __name__=="__main__":
    print(totalNQueens(4))
    print(totalNQueens(6))

File: CPP/Lesson_172_n_queens.cpp
// =============================================================
// MIT License | @analyticswithharry2026
// GitHub  : https://github.com/analyticswithharry
// YouTube : Analytics with Harry
// =============================================================
// Lesson     : 172 -- N Queens
// Category   : Backtracking
// Difficulty : Hard
// Study Plan : Day 86
// =============================================================
//
// QUESTION:
//   Place n queens on n×n board so none attack each other; return the count of distinct solutions.
// =============================================================
#include <vector>
#include <string>
#include <iostream>
#include <stack>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <map>
#include <set>
#include <algorithm>
#include <climits>
#include <numeric>
#include <functional>
#include <cmath>
using namespace std;
int CNT,N; set<int> COLS,D1,D2;
void bt(int r){if(r==N){CNT++;return;}for(int c=0;c<N;c++){if(COLS.count(c)||D1.count(r-c)||D2.count(r+c))continue;COLS.insert(c);D1.insert(r-c);D2.insert(r+c);bt(r+1);COLS.erase(c);D1.erase(r-c);D2.erase(r+c);}}
int totalNQueens(int n){CNT=0;N=n;COLS.clear();D1.clear();D2.clear();bt(0);return CNT;}
int main(){cout<<totalNQueens(4)<<"\n"<<totalNQueens(6)<<"\n";}
